### PR TITLE
1860: Make DistanceGraph class

### PR DESCRIPTION
**Pull request by roseundy**
Move game-specific distance_graph out of 1860 and into its own class.

### CODE REVIEW

File: lib/engine/distance_graph.rb [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+      distance: nil,
+      node_distances: {},
+      path_distances: {},
+      corporation: nil,
+      counter: Hash.new(0),
+      &block
+    )
+      return if smaller_or_equal_distance?(node_distances[node], distance)
+
+      node_distances[node] = merge_distance(node_distances[node], distance)
```
Comment (tobymao) [inline lib/engine/distance_graph.rb:67]: this is expensive, can it be avoided?

File: lib/engine/distance_graph.rb [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      elsif node.city?
+        distance[:city] += 1
+      elsif node.town? && !node.halt?
+        distance[:town] += 1
+      end
+
+      node.paths.each do |node_path|
+        next if smaller_or_equal_distance?(path_distances[node_path], distance)
+
+        node_path.walk(
```
Comment (tobymao) [inline lib/engine/distance_graph.rb:81]: one line

File: lib/engine/distance_graph.rb [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+          counter: counter,
+        ) do |path, _vp, ct|
+          path_distances[path] = merge_distance(path_distances[path], distance)
```
Comment (tobymao) [inline lib/engine/distance_graph.rb:84]: expensive

File: lib/engine/distance_graph.rb [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    def merge_distance(a, b)
+      return b.dup unless a
+
+      a.map { |k, v| [k, [v, b[k]].min] }.to_h
+    end
+
+    def node_walk(
+      node,
+      distance: nil,
+      node_distances: {},
```
Comment (tobymao) [inline lib/engine/distance_graph.rb:59]: why make these default? do you ever pass it whit nil?

File: lib/engine/distance_graph.rb [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+      tokens.each do |node|
+        node_walk(
+          node,
+          distance: @separate_node_types ? { city: 0, town: 0 } : { node: 0 },
+          node_distances: n_distances,
+          path_distances: p_distances,
+          corporation: corporation,
+        ) do |path, dist|
+          hex = path.hex
+          h_distances[hex] = merge_distance(h_distances[hex], dist)
```
Comment (tobymao) [inline lib/engine/distance_graph.rb:130]: you're calling this a lot...

File: lib/engine/distance_graph.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/distance_graph.rb:81]: OK

File: lib/engine/distance_graph.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/distance_graph.rb:59]: Fixed.

File: lib/engine/distance_graph.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
+
+module Engine
+  class DistanceGraph
+    def initialize(game, **opts)
+      @game = game
+      @node_distances = {}
+      @path_distances = {}
+      @hex_distances = {}
+      @separate_node_types = opts[:separate_node_types] || false
```
Comment (tobymao) [inline lib/engine/distance_graph.rb:10]: don't need the false here

Comment (roseundy) [reply lib/engine/distance_graph.rb:10]: OK

File: lib/engine/distance_graph.rb [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+      tokens
+    end
+
+    def smaller_or_equal_distance?(a, b)
+      a&.all? { |k, v| v <= b[k] }
+    end
+
+    def merge_distance(dict, key, b)
+      if (a = dict[key])
+        a.keys.each { |k| a[k] = [a[k], b[k]].min }
```
[the remark this answers is not in the document]
Comment (roseundy) [reply lib/engine/distance_graph.rb:52]: OK

**Pull request merged**